### ISSUE TITLE
`simp` doesn't always synthesize instances depending on `outParam`s

### ISSUE THREAD

**Issue by Vierkantor** [leanprover/lean4#3107]
### Prerequisites

* [x] Put an X between the brackets on this line if you have done all of the following:
    * Check that your issue is not already [filed](https://github.com/leanprover/lean4/issues).
    * Reduce the issue to a minimal, self-contained, reproducible test case. Avoid dependencies to mathlib4 or std4.

### Description

In `simp` calls, finding an instance that depends on an `outParam` of an instance later in the parameter list is not reliable. The following minimized example comes from leanprover-community/mathlib4#8386:

```lean
class Semiring (R : Type)

class Module (R M : Type) [Semiring R] :=
  (foo : M)

class FunLike (F : Type) (α β : outParam Type) :=
  (toFun : F → α → β)

class SemilinearMapClass (F : Type) (R S M N : outParam Type) [Semiring R] [Semiring S]
    [Module R M] [Module S N] [FunLike F M N] :=
  (map_smul : ∀ (f : F) (x : M), FunLike.toFun f x = Module.foo S)

set_option trace.Meta.synthInstance true in
set_option trace.Meta.Tactic.simp true in
example (F R S M N : Type) [Semiring R] [Semiring S] [Module R M] [Module S N] [FunLike F M N]
    [SemilinearMapClass F R S M N] (f : F) (x : M) :
    FunLike.toFun f x = Module.foo S :=
  by simp only [SemilinearMapClass.map_smul] -- simp made no progress
/-
Trace output:
[Meta.synthInstance] 💥 Semiring ?R ▶

[Meta.Tactic.simp.discharge] @SemilinearMapClass.map_smul, failed to synthesize instance
      Semiring ?R

[Meta.Tactic.simp.unify] @eq_self:1000, failed to unify
      ?a = ?a
    with
      FunLike.toFun f x = Module.foo S

-/
```

The equivalent Lean 3 code succeeds:
```lean
class Semiring (R : Type)

class Module (R M : Type) [Semiring R] :=
  (foo : M)

class FunLike (F : Type) (α β : out_param Type) :=
  (toFun : F → α → β)

class SemilinearMapClass (F : Type) (R S M N : out_param Type) [Semiring R] [Semiring S]
    [Module R M] [Module S N] [FunLike F M N] :=
  (map_smul : ∀ (f : F) (x : M), FunLike.toFun f x = Module.foo S)

example (F R S M N : Type) [Semiring R] [Semiring S] [Module R M] [Module S N] [FunLike F M N]
    [SemilinearMapClass F R S M N] (f : F) (x : M) :
    FunLike.toFun f x = Module.foo S :=
  by simp only [SemilinearMapClass.map_smul] -- simp succeeds
```

### Context

[For performance reasons](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/.60example.20.28p.20.3A.20P.29.20.3A.20Q.20.3A.3D.20p.60.20takes.200.2E25s.20to.20fail!) I would like to change the definition of "morphism classes", including `SemilinearMapClass` in Mathlib: previously we wrote `extends FunLike F α β`, which we want to change to a parameter `[FunLike F α β]`. All morphism classes can succesfully be changed, apart from `SemilinearMapClass`, because currently `simp` does not reliably apply the `map_smul` lemma.

### Steps to Reproduce

See minimized example code. The crucial part appears to be a `simp` lemma taking an instance parameter, which depends on an `outParam` of a class that does not appear in the `simp` goal.

**Expected behavior:** `simp` finds the `SemilinearMapClass` instance, assigns `R` and `S` using `outParam`s, and uses this to synthesize `Semiring R` `Semiring S` instances, like in Lean 3.

**Actual behavior:** `simp` fails when it cannot synthesize `Semiring` instances, even though it could make progress.

### Versions

```lean
#eval Lean.versionString
-- "4.5.0, commit 5ac244a932e52a14ec69ac20d41c5e268ffa1d62"
```

```bash
$ uname -a
Linux McChouffe 6.1.62-1-lts #1 SMP PREEMPT_DYNAMIC Thu, 09 Nov 2023 17:21:17 +0000 x86_64 GNU/Linux
```

### Impact

Add :+1: to [issues you consider important](https://github.com/leanprover/lean4/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc). If others are impacted by this issue, please ask them to add :+1: to it.

**Comment by kmill** [comment]
For some more context, @Vierkantor implemented in https://github.com/leanprover-community/lean/pull/659 an instance synthesis loop for simp lemma arguments, which is similar in spirit to the Lean 4 main elaborator's instance synthesis loop.

An analogous change to Lean 4 would be to have `Lean.Meta.Simp.synthesizeArgs` maintain a work list of instance metavariables, and make it so it fails only if no progress can be made or if there is a true synthesis failure. One design complication is deciding at what stage simp should discharge Prop arguments (maybe they could also be on the work list, where, like tactic metavariables, they are deferred if their type still has metavariables?)

It would be nice to make it so that, however this algorithm is changed, the algorithm for how `rw` synthesizes its arguments follows suit.

**Comment by Vierkantor** [comment]
I should also point out that `by rw [SemilinearMapClass.h]` and `SemilinearMapClass.h _ _` successfully prove the goal, but `by apply SemilinearMapClass.h` does not.

**Comment by negiizhao** [comment]
Has this already been resolved?

**Comment by semorrison** [comment]
This does seem to be fix on v4.8.0-rc1!
